src/main/java/com/spotify/google/cloud/pubsub/client/Puller.java

```diff
@@ -146,25 +148,32 @@ public long pullIntervalMillis() {
 
   private void pull() {
     while (outstandingRequests.get() < concurrency &&
-           outstandingMessages.get() < maxOutstandingMessages) {
+           outstandingMessages.get() < maxOutstandingMessages &&
+           hasData.get()) {
       pullBatch();
     }
   }
 
   private void pullBatch() {
     outstandingRequests.incrementAndGet();
 
-    pubsub.pull(project, subscription, false, batchSize)
+    pubsub.pull(project, subscription, true, batchSize)
         .whenComplete((messages, ex) -> {
 
           outstandingRequests.decrementAndGet();
-
           // Bail if pull failed
           if (ex != null) {
             log.error("Pull failed", ex);
             return;
           }
 
+          if (messages.size() == 0) {
```

Comment:
It feels like this would block the pulling forever.

When you set it on false once you never go in again.

Also this flag is set on all the threads that are using to pull, so I don't think this is something that would work.